File: common/src/test/java/DahRealPanda/plugins/colonyweb/map/ColonyMapTest.java
package DahRealPanda.plugins.colonyweb.map;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * The colony map raster.
 *
 * <p>The behaviour worth protecting is what happens when a colony outgrows its map: the new
 * raster inherits everything the old one drew, because the chunks it would otherwise blank out
 * are exactly the ones the server no longer has loaded and cannot redraw.</p>
 */
class ColonyMapTest {

    private static ColonyMap map(int minX, int minZ, int size) {
        return new ColonyMap(minX, minZ, size, size);
    }

    @Nested
    @DisplayName("geometry")
    class Geometry {

        @Test
        @DisplayName("chunk bounds are derived from the block bounds")
        void chunkBounds() {
            ColonyMap colonyMap = map(-64, -64, 128);

            assertEquals(-4, colonyMap.chunkX);
            assertEquals(-4, colonyMap.chunkZ);
            assertEquals(8, colonyMap.chunkCols);
            assertEquals(8, colonyMap.chunkRows);
            assertEquals(64, colonyMap.chunkCount());
        }

        @Test
        @DisplayName("the rasters are sized for one pixel per block")
        void rasterSize() {
            ColonyMap colonyMap = map(0, 0, 64);

            assertEquals(64 * 64, colonyMap.rgb.length);
            assertEquals(64 * 64, colonyMap.top.length);
        }

        @Test
        @DisplayName("a fresh map has drawn nothing")
        void startsEmpty() {
            ColonyMap colonyMap = map(0, 0, 64);

            assertEquals(0, colonyMap.mapped);
            assertEquals(0, colonyMap.cursor);
            for (long stamp : colonyMap.chunkStamp) {
                assertEquals(0, stamp);
            }
        }
    }

    @Nested
    @DisplayName("covers")
    class Covers {

        @Test
        @DisplayName("an area inside the map is covered")
        void inside() {
            assertTrue(map(0, 0, 128).covers(16, 16, 64, 64));
        }

        @Test
        @DisplayName("the exact bounds are covered")
        void exactBounds() {
            assertTrue(map(0, 0, 128).covers(0, 0, 128, 128));
        }

        @Test
        @DisplayName("an area poking out on any side is not covered")
        void outside() {
            ColonyMap colonyMap = map(0, 0, 128);

            assertFalse(colonyMap.covers(-1, 0, 64, 64), "past the west edge");
            assertFalse(colonyMap.covers(0, -1, 64, 64), "past the north edge");
            assertFalse(colonyMap.covers(0, 0, 129, 64), "past the east edge");
            assertFalse(colonyMap.covers(0, 0, 64, 129), "past the south edge");
        }

        @Test
        @DisplayName("negative coordinates work the same way")
        void negativeOrigin() {
            ColonyMap colonyMap = map(-128, -128, 128);

            assertTrue(colonyMap.covers(-128, -128, -64, -64));
            assertFalse(colonyMap.covers(-129, -128, -64, -64));
        }
    }

    @Nested
    @DisplayName("inheriting from an outgrown map")
    class Inheriting {

        @Test
        @DisplayName("pixels the old map drew are copied into the overlap")
        void copiesPixels() {
            ColonyMap old = map(0, 0, 32);
            old.rgb[0] = 0x123456;
            old.top[0] = 70;

            ColonyMap grown = map(0, 0, 64);
            grown.inherit(old);

            assertEquals(0x123456, grown.rgb[0]);
            assertEquals(70, grown.top[0]);
        }

        @Test
        @DisplayName("pixels land at the right place when the origin moves")
        void reindexesOnShift() {
            ColonyMap old = map(0, 0, 32);
            // Block (16, 16) in the old map.
            old.rgb[16 * 32 + 16] = 0xABCDEF;

            // The new map starts 16 blocks further west and north, so that same block moves.
            ColonyMap grown = new ColonyMap(-16, -16, 64, 64);
            grown.inherit(old);

            assertEquals(0xABCDEF, grown.rgb[(16 + 16) * 64 + (16 + 16)]);
        }

        @Test
        @DisplayName("chunk timestamps and the drawn count carry over")
        void carriesChunkStamps() {
            ColonyMap old = map(0, 0, 32);
            old.chunkStamp[0] = 1234L;
            old.mapped = 1;

            ColonyMap grown = map(0, 0, 64);
            grown.inherit(old);

            assertEquals(1234L, grown.chunkStamp[0]);
            assertEquals(1, grown.mapped);
        }

        @Test
        @DisplayName("chunks outside the old map stay undrawn")
        void newAreaStaysUndrawn() {
            ColonyMap old = map(0, 0, 32);
            for (int i = 0; i < old.chunkStamp.length; i++) {
                old.chunkStamp[i] = 1234L;
            }
            old.mapped = old.chunkStamp.length;

            ColonyMap grown = map(0, 0, 64);
            grown.inherit(old);

            assertEquals(old.chunkStamp.length, grown.mapped,
                    "only the chunks the old map actually had should count as drawn");
            assertTrue(grown.mapped < grown.chunkCount(), "the new area is still blank");
        }

        @Test
        @DisplayName("two maps that do not overlap copy nothing")
        void disjointMaps() {
            ColonyMap old = map(0, 0, 32);
            old.rgb[0] = 0x123456;
            old.chunkStamp[0] = 1234L;

            ColonyMap elsewhere = map(1024, 1024, 32);
            elsewhere.inherit(old);

            assertEquals(0, elsewhere.rgb[0]);
            assertEquals(0, elsewhere.mapped);
        }

        @Test
        @DisplayName("the version keeps climbing so browsers do not serve a cached image")
        void keepsVersion() {
            ColonyMap old = map(0, 0, 32);
            old.version = 7;
            old.renderedAt = 999L;
            old.png = new byte[]{1, 2, 3};

            ColonyMap grown = map(0, 0, 64);
            grown.inherit(old);

            assertEquals(7, grown.version);
            assertEquals(999L, grown.renderedAt);
            assertNotEquals(old.png, grown.png,
                    "the old PNG is the wrong size for the new bounds and must not be reused");
        }
    }

    @Nested
    @DisplayName("draw order")
    class DrawOrder {

        @Test
        @DisplayName("every chunk appears exactly once")
        void coversEveryChunk() {
            ColonyMap colonyMap = map(0, 0, 64);

            boolean[] seen = new boolean[colonyMap.chunkCount()];
            for (int index : colonyMap.order) {
                assertFalse(seen[index], "chunk " + index + " listed twice");
                seen[index] = true;
            }
            for (boolean visited : seen) {
                assertTrue(visited, "every chunk should be scheduled");
            }
        }

        @Test
        @DisplayName("the colony itself is drawn before its outskirts")
        void centreFirst() {
            ColonyMap colonyMap = map(0, 0, 64);
            int cols = colonyMap.chunkCols;
            double mid = (cols - 1) / 2.0;

            double first = distanceFromCentre(colonyMap.order[0], cols, mid);
            double last = distanceFromCentre(colonyMap.order[colonyMap.order.length - 1], cols, mid);

            assertTrue(first < last, "the first chunk drawn should be nearer the middle");
        }

        @Test
        @DisplayName("the order never moves outward and then back in")
        void monotonic() {
            ColonyMap colonyMap = map(0, 0, 128);
            int cols = colonyMap.chunkCols;
            double mid = (cols - 1) / 2.0;

            double previous = -1;
            for (int index : colonyMap.order) {
                double distance = distanceFromCentre(index, cols, mid);
                assertTrue(distance >= previous, "draw order jumped back towards the centre");
                previous = distance;
            }
        }

        private double distanceFromCentre(int index, int cols, double mid) {
            double dx = (index % cols) - mid;
            double dz = (index / cols) - mid;
            return dx * dx + dz * dz;
        }
    }
}

File: common/src/test/kotlin/DahRealPanda/plugins/colonyweb/renderer/ColonyMapTest.kt
package DahRealPanda.plugins.colonyweb.renderer

import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe

/**
 * The colony map is one pixel per block, drawn a chunk at a time and re-cut whenever the colony
 * grows. All of that rests on the coordinate arithmetic here — which has to hold up west and north
 * of the origin, where a colony is just as likely to be as anywhere else.
 */
class ColonyMapTest : DescribeSpec({

    describe("geometry") {
        it("derives the chunk grid from the block bounds") {
            val map = ColonyMap(minX = 0, minZ = 0, width = 128, height = 64)

            map.chunkX shouldBe 0
            map.chunkZ shouldBe 0
            map.chunkCols shouldBe 8
            map.chunkRows shouldBe 4
            map.chunkCount() shouldBe 32
        }

        it("allocates one pixel per block") {
            val map = ColonyMap(minX = 0, minZ = 0, width = 128, height = 64)

            map.rgb.size shouldBe 128 * 64
            map.top.size shouldBe 128 * 64
        }

        // A colony west or north of the origin has negative bounds, and an integer division would
        // round those towards zero — putting the map's first chunk one chunk too far east.
        it("rounds negative bounds down to the chunk that contains them") {
            val map = ColonyMap(minX = -128, minZ = -64, width = 128, height = 64)

            map.chunkX shouldBe -8
            map.chunkZ shouldBe -4
        }

        it("rounds a bound that is not on a chunk boundary down to the chunk containing it") {
            ColonyMap(minX = 17, minZ = -1, width = 32, height = 32).let { map ->
                map.chunkX shouldBe 1
                map.chunkZ shouldBe -1
            }
        }

        it("starts with nothing drawn") {
            val map = ColonyMap(minX = 0, minZ = 0, width = 64, height = 64)

            map.mapped shouldBe 0
            map.cursor shouldBe 0
            map.version shouldBe 0
            map.renderedAt shouldBe 0L
            map.png shouldBe null
        }

        // Chunks are drawn from the middle outwards, because the town hall is at the centre and
        // that is the part of the map a player looks at first.
        it("draws from the centre outwards") {
            val map = ColonyMap(minX = 0, minZ = 0, width = 48, height = 48)

            map.order.size shouldBe 9
            map.order.first() shouldBe 4
        }
    }

    describe("covers") {
        val map = ColonyMap(minX = 0, minZ = 0, width = 128, height = 128)

        it("covers a region well inside its bounds") {
            map.covers(10, 10, 100, 100) shouldBe true
        }

        it("covers a region that reaches exactly to its bounds") {
            map.covers(0, 0, 128, 128) shouldBe true
        }

        it("does not cover a region that runs off any edge") {
            map.covers(-1, 0, 100, 100) shouldBe false
            map.covers(0, -1, 100, 100) shouldBe false
            map.covers(0, 0, 129, 100) shouldBe false
            map.covers(0, 0, 100, 129) shouldBe false
        }

        it("does not cover a region that misses it entirely") {
            map.covers(1000, 1000, 1100, 1100) shouldBe false
        }

        it("works the same for a map west and north of the origin") {
            val negative = ColonyMap(minX = -256, minZ = -256, width = 128, height = 128)

            negative.covers(-256, -256, -128, -128) shouldBe true
            negative.covers(-257, -256, -128, -128) shouldBe false
        }
    }

    // When a colony grows the map is re-cut around the new bounds. Redrawing it from scratch would
    // blank the dashboard for as long as it takes to walk the chunks again, so the new map takes
    // over the pixels and the per-chunk timestamps of the old one.
    describe("inheriting from the previous map") {
        it("carries over the pixels of the overlapping area") {
            val old = ColonyMap(minX = 0, minZ = 0, width = 32, height = 32)
            old.rgb[0] = 0x00FF00
            old.rgb[31 * 32 + 31] = 0x0000FF

            val grown = ColonyMap(minX = 0, minZ = 0, width = 64, height = 64)
            grown.inherit(old)

            grown.rgb[0] shouldBe 0x00FF00
            grown.rgb[31 * 64 + 31] shouldBe 0x0000FF
        }

        it("carries over which chunks were already drawn") {
            val old = ColonyMap(minX = 0, minZ = 0, width = 32, height = 32)
            old.chunkStamp[0] = 12345L
            old.version = 7
            old.renderedAt = 999L

            val grown = ColonyMap(minX = 0, minZ = 0, width = 64, height = 64)
            grown.inherit(old)

            grown.chunkStamp[0] shouldBe 12345L
            grown.mapped shouldBe 1
            grown.version shouldBe 7
            grown.renderedAt shouldBe 999L
        }

        it("leaves the newly added area undrawn") {
            val old = ColonyMap(minX = 0, minZ = 0, width = 32, height = 32)
            old.chunkStamp.fill(1L)

            val grown = ColonyMap(minX = 0, minZ = 0, width = 64, height = 64)
            grown.inherit(old)

            grown.mapped shouldBe 4
            grown.chunkCount() shouldBe 16
        }

        it("carries over only the part that still overlaps when the map shifts") {
            val old = ColonyMap(minX = 0, minZ = 0, width = 64, height = 64)
            old.rgb[0] = 0x00FF00
            old.rgb[32 * 64 + 32] = 0x0000FF

            val shifted = ColonyMap(minX = 32, minZ = 32, width = 64, height = 64)
            shifted.inherit(old)

            shifted.rgb[0] shouldBe 0x0000FF
        }

        it("inherits nothing from a map that no longer overlaps at all") {
            val old = ColonyMap(minX = 0, minZ = 0, width = 32, height = 32)
            old.rgb.fill(0x00FF00)
            old.chunkStamp.fill(1L)

            val elsewhere = ColonyMap(minX = 1024, minZ = 1024, width = 32, height = 32)
            elsewhere.inherit(old)

            elsewhere.mapped shouldBe 0
            elsewhere.rgb.all { it == 0 } shouldBe true
        }
    }
})
